src/lsm/lsm_cursor.c

```diff
@@ -902,6 +944,92 @@ err:	__clsm_leave(clsm);
 }
 
 /*
+ * __clsm_random_chunk --
+ *	Pick a chunk at random, weighted by the size of all chunks. Weighting
+ * proportional to documents avoids biasing towards small chunks. Then return
+ * the cursor on the chunk we have picked.
+ */
+static int
+__clsm_random_chunk(WT_SESSION_IMPL *session,
+    WT_CURSOR_LSM *clsm, WT_CURSOR **cursor)
+{
+	uint64_t checked_docs, i, rand_doc, total_docs;
+
+	/*
+	 * If the tree is empty we cannot do a random lookup, so return a
+	 * WT_NOTFOUND.
+	 */
+	if (clsm->nchunks == 0)
+		return (WT_NOTFOUND);
+	for (total_docs = i = 0; i < clsm->nchunks; i++) {
+		total_docs += clsm->iters[i]->count;
+	}
+	if (total_docs == 0)
+		return (WT_NOTFOUND);
+
+	rand_doc = __wt_random(&session->rnd) % total_docs;
+
+	for (checked_docs = i = 0; i < clsm->nchunks; i++) {
+		checked_docs += clsm->iters[i]->count;
+		if (rand_doc <= checked_docs) {
+			*cursor = clsm->iters[i]->cursor;
+			break;
+		}
+	}
+	return (0);
+}
+
+/*
+ * __clsm_next_random --
+ *	WT_CURSOR->next method for the LSM cursor type when configured with
+ * next_random.
+ */
+static int
+__clsm_next_random(WT_CURSOR *cursor)
+{
+	WT_CURSOR_LSM *clsm;
+	WT_CURSOR *c;
+	WT_DECL_RET;
+	WT_SESSION_IMPL *session;
+	int exact;
+
+	c = NULL;
+	clsm = (WT_CURSOR_LSM *)cursor;
+
+	CURSOR_API_CALL(cursor, session, next, NULL);
+	WT_CURSOR_NOVALUE(cursor);
+	WT_ERR(__clsm_enter(clsm, false, false));
+
+	do {
+		WT_ERR(__clsm_random_chunk(session, clsm, &c));
+		ret = __wt_curfile_next_random(c);
+		/*
+		 * Sometimes we may run on an empty chunk or for some reason be
+		 * unable to get a random doc. That's okay and we retry.
+		 */
+		if (ret == 0) {
+			F_SET(cursor, WT_CURSTD_KEY_INT);
+			WT_ERR(c->get_key(c, &cursor->key));
+			/*
+			 * Search near the current key to resolve any tombstones
+			 * and position to a valid document. If we see a
+			 * WT_NOTFOUND here that is valid, as the tree has no
+			 * documents visible to us.
+			 */
+			WT_ERR(__clsm_search_near(cursor, &exact));
+		}
+	} while (ret == WT_NOTFOUND);
```

Comment:
It might be better to restructure this loop as a `for (;;)` with explicit `continue` / `break`.  The fact that `WT_NOTFOUND` sometimes continues and sometimes causes a `goto err` is unnecessarily subtle.